### ISSUE TITLE
Expose window geometry saved for awful.client.floating via API

### ISSUE THREAD

**Issue by vdrandom** [awesomeWM/awesome#172]
Saving windows' geometry in awful.layout.suit.floating so that it is restored when switching between layouts can be easily implemented in the config, but awful.client.floating uses its own way of saving and restoring window geometry. It would have been nice to get access to it. First, because it saves initial window geometry when program is launched in any other layout than awful.layout.suit.floating and second, because it seems logical to save window geometry between awful.layout.suit.floating and awful.client.floating.

**Comment by blueyed** [comment]
It uses floating_geometry internally, which gets initialized and then updated for floating clients.
This can be accessed through `c.floating_geometry`.
If you want to keep the initial value, you would have to save it yourself through the `new` signal, too.

See https://github.com/awesomeWM/awesome/pull/168, which contains a fix for handling this properly for (partially) maximized clients.

I've noticed that the property only gets stored for floating clients, but not in general for clients on the floating layout.  This could be changed (fixed?) easily.

But then the question is how a client should behave when it is not floating, and the layout gets switch from floating to a a tiling one and back to floating: should the clients be restored to where they were on the floating layout? (I guess so, and that's probably what this issue is about mainly)

**Comment by vdrandom** [comment]
The thing does not seem to be accessible by calling ```c.floating_geometry``` table, but works fine with ```awful.client.property.set(c, 'floating_geometry', c:geometry())``` and ```awful.client.property.get(c, 'floating_geometry')```.

This is exactly what I was looking for, thank you. It feels like the API documentation is a bit lacking in terms of properties.

> I've noticed that the property only gets stored for floating clients, but not in general for clients on the floating layout. This could be changed (fixed?) easily.
>
> But then the question is how a client should behave when it is not floating, and the layout gets switch from floating to a a tiling one and back to floating: should the clients be restored to where they were on the floating layout? (I guess so, and that's probably what this issue is about mainly)

Yes, the only thing missing is the logic to place windows properly if it is the first time they enter floating mode (```floating_geometry``` property starts with ```{ x = 0, y = 0 }``` by default.)

The first idea is to store the initial ```floating_geometry``` somewhere and then, when switching to ```awful.layout.suit.floating```, compare the value against current ```floating_geometry```. Another option would be to make a new ```awful.client``` property and set it to ```true``` if the window position was changed manually or on spawn.

Here is the code snippet I am using right now.
```lua
tag.connect_signal(
	'property::layout',
	function(t)
		for k, c in ipairs(t:clients()) do
			if awful.layout.get(mouse.screen) == awful.layout.suit.floating then
				c:geometry(awful.client.property.get(c, 'floating_geometry'))
			end
		end
	end
)

client.connect_signal(
	'manage',
	function(c)
		if awful.layout.get(mouse.screen) == awful.layout.suit.floating then
			awful.client.property.set(c, 'floating_geometry', c:geometry())
		end
	end
)

client.connect_signal(
	'property::geometry',
	function(c)
		if awful.layout.get(mouse.screen) == awful.layout.suit.floating then
			awful.client.property.set(c, 'floating_geometry', c:geometry())
		end
	end
)
```

**Comment by blueyed** [comment]
Yes, `awful.client.property.get(c, 'floating_geometry')` is correct. Sorry.

I've pushed another commit to https://github.com/blueyed/awesome/commit/bc7936532c31fecd76be67a796612fb238c772a0, which looks not as clean as your approach.

It's a bit of a mess, and probably fits better into the ewmh module..

**Comment by blueyed** [comment]
@vdrandom 
I'd appreciate feedback on #168, which should fix this issue for you.

**Comment by vdrandom** [comment]
Works as intended.
Also would be nice to have it store sizes when moving windows between tags with and without floating layout. If you move a window from tag with floating layout to another tag with non-floating (like, maximized or tiled), the new window sizes are applied, and when you switch back to floating or move the window to tag with floating layout, it keeps the size. It can be switched back to the floating size in floating window mode, so I think we should restore that size on moving back to a tag with floating layout.

**Comment by sammoth** [comment]
If anyone sees this, the following seems to work including when moving clients between tags and showing/hiding extra tags etc.

```
tag.connect_signal(
    'property::layout',
    function(t)
        for k, c in ipairs(t:clients()) do
            if awful.layout.get(mouse.screen) == awful.layout.suit.floating then
                c:geometry(awful.client.property.get(c, 'floating_geometry'))
            end
        end
    end
)

tag.connect_signal(
    'property::selected',
    function(t)
        for k, c in ipairs(t:clients()) do
            if awful.layout.get(mouse.screen) == awful.layout.suit.floating then
                c:geometry(awful.client.property.get(c, 'floating_geometry'))
            end
        end
    end
)

client.connect_signal(
    'request::manage',
    function(c)
        if awful.layout.get(mouse.screen) == awful.layout.suit.floating then
            awful.client.property.set(c, 'floating_geometry', c:geometry())
        end
    end
)

client.connect_signal(
    'property::geometry',
    function(c)
        if awful.layout.get(mouse.screen) == awful.layout.suit.floating then
            awful.client.property.set(c, 'floating_geometry', c:geometry())
        end
    end
)
```
